### ISSUE TITLE
Base code works, but generated code segfaults

### ISSUE THREAD

**Issue by massimiliano-leoni** [OP2/OP2-Common#77]
Hi everyone, I recently started learning OP2 for my internship in RR, and I am having some trouble having it work properly.

I have a simple code that I wrote and wrapped [2D heat equation, finite differences] that works fine in sequential mode with op2 wrappers, but the openmp version generated by op2.py segfaults when calling `op_plan_core` in one of the kernels [not the first that is executed].
As far as I understood from the documentation, if the base code [sequential + wrappers] works fine, the generated code is supposed to run out-of-the-box [please correct me if wrong], so I thought I filed this issue report.

I can provide more details, just tell me what you need.

**Comment by gihanmudalige** [comment]
Hi Massimiliano, could you share the code you were trying to code generate with us. I can have a look.

**Comment by massimiliano-leoni** [comment]
Sure, and thank you for the quick response :)
I took the time to add basic comments and write a README.

Here's the file. I had to put it on my dropbox as apparently github doesn't allow to attach a generic file.
https://www.dropbox.com/s/i6pq07fkshqcq4l/heat.tar.gz?dl=0

**Comment by massimiliano-leoni** [comment]
Hi, I don't want to rush anybody, but I was just wondering if it makes sense that I look for a bug in the code that I wrote or if it's likely something to tweak in op2.

**Comment by gihanmudalige** [comment]
Hi Massimiliano,
I will have a look at this tomorrow. I am having some trouble logging in to our systems today. 
From just a visual inspection, doing this type of direct memcopy from OP2 space to user space is not a good idea. 
Could you let me know whether the crash occours  in the very first time iteration (i.e. k = 0), just before the memcopy ?

**Comment by massimiliano-leoni** [comment]
Hi Gihan,
the memcopy part was also my first suspect, so I promptly substituted it with an "update" kernel routine that simply copies the values explicitly.
Anyway, the crash is occurring before that part, at the very first iteration, so this change did not affect it.
The output of the code is
    let's do this
    new execution plan #0 for kernel computeValue
    Segmentation fault (core dumped)

**Comment by gihanmudalige** [comment]
Hi Massimiliano, I am looking through your code now and trying to understand your application. Your neighbnodes array has negative values. This is suppose to be the mapping table that provides the indirection. Now I am wondering how it is possible to access a negative index.

**Comment by massimiliano-leoni** [comment]
Hi Gihan,
In my mind, the `neighbnodes` array contains 5 integers for each node, indicating the indexes of the nodes that are its neighbours in the FD stencil.
The neighbnodes with negative values are the ones corresponding to the lower boundary nodes. Although in principle wrong, these values are never actually used because in the `computeValue` routine the boundary nodes are automatically ruled out by the `isBoundary` function, that prevents the program to go into the part of the function where they would be used.
The first node for which the neighbnodes array is actually used is the Nx+2/th, and its neighbour nodes should all have positive indexes.

**Comment by gihanmudalige** [comment]
Hi Massimiliano, 
I see what you mean. However OP2 should have checked whether the indices in a mapping table points to valid set element indices. This check used to be there but had been commented out during the development of MPI. 

I have just included this check in a branch (app/heat) and have pushed it to the OP2 github repo. Could you use this branch to fix your application. I.e. the sequential version should now tell you there is an issue with your mapping table. 

For other parallelisations especially MPI, it is not good to have mapping tables with negative indices. This will break the distributed memory partitioning. 

I suspect it is a similar issue that's causing the plan function to fail.  If you could update your application and if OP2's plan function still fails I can re-investigate.

**Comment by massimiliano-leoni** [comment]
Ok, I understand the problem here. I set all the negative indexes to zero using the `isBoundary` function and now it seems to work properly. Excellent!
I tried to checkout the new branch but it doesn't compile b/c of some unused variable, and I wasn't able to disable -Werror, so I tried in the master branch, thus not triggering the check that caused the segfault.
I can try to do that tomorrow if you want.
